src/hotspot/share/opto/node.cpp

```diff
@@ -1696,6 +1696,670 @@ bool Node::add_to_worklist(Node* n, Node_List* worklist, Arena* old_arena, Vecto
+    }
+  }
+  if (!_traverse_inputs && !_traverse_outputs) {
+    _traverse_inputs = true;
+  }
+  if (_filter_visit.is_empty()) {
+    _filter_visit.set_all();
+  }
+  Compile* C = Compile::current();
+  _print_old &= (C->matcher() != nullptr); // only show old if there are new
+  _print_blocks &= (C->cfg() != nullptr); // only show blocks if available
+  return true;
+}
+
+bool PrintBFS::filter_category(Node* n, Filter& filter) {
+  const Type* t = n->bottom_type();
+  switch (t->category()) {
+    case Type::Category::Data:
+      return filter._data;
+    case Type::Category::Memory:
+      return filter._memory;
+    case Type::Category::Mixed:
+      return filter._mixed;
+    case Type::Category::Control:
+      return filter._control;
+    case Type::Category::Other:
+      return filter._other;
+    case Type::Category::Undef:
+      n->dump();
+      assert(false, "category undef ??");
+    default:
+      n->dump();
+      assert(false, "not covered");
+  }
+  return false;
+}
+
+void PrintBFS::DumpConfigColored::pre_dump(outputStream* st, const Node* n) {
+  if (!_bfs->_use_color) {
+    return;
+  }
+  Info* info = _bfs->find_info(n);
+  if (info == nullptr || !info->is_marked()) {
+    return;
+  }
+
+  const Type* t = n->bottom_type();
+  switch (t->category()) {
+    case Type::Category::Data:
+      st->print("\u001b[34m");
+      break;
+    case Type::Category::Memory:
+      st->print("\u001b[32m");
+      break;
+    case Type::Category::Mixed:
+      st->print("\u001b[35m");
+      break;
+    case Type::Category::Control:
+      st->print("\u001b[31m");
+      break;
+    case Type::Category::Other:
+      st->print("\u001b[33m");
+      break;
+    case Type::Category::Undef:
+      n->dump();
+      assert(false, "category undef ??");
+      break;
+    default:
+      n->dump();
+      assert(false, "not covered");
+      break;
+  }
+}
+
+void PrintBFS::DumpConfigColored::post_dump(outputStream* st) {
+  if (!_bfs->_use_color) {
+    return;
+  }
+  st->print("\u001b[0m"); // white
+}
+
+Node* PrintBFS::old_node(Node* n) {
+  Compile* C = Compile::current();
+  if (C->matcher() == nullptr || !C->node_arena()->contains(n)) {
+    return (Node*)nullptr;
+  } else {
+    return C->matcher()->find_old_node(n);
+  }
+}
+
+void PrintBFS::print_node_idx(Node* n) {
+  Compile* C = Compile::current();
+  char buf[30];
+  if (n == nullptr) {
+    sprintf(buf,"_");           // null
+  } else if (C->node_arena()->contains(n)) {
+    sprintf(buf, "%d", n->_idx);  // new node
+  } else {
+    sprintf(buf, "o%d", n->_idx); // old node
+  }
+  tty->print("%6s", buf);
+}
+
+void PrintBFS::print_node_block(Node* n) {
+  Compile* C = Compile::current();
+  Block* b = C->node_arena()->contains(n)
+             ? C->cfg()->get_block_for_node(n)
+             : nullptr; // guard against old nodes
+  if (b == nullptr) {
+    tty->print("     _");
+    tty->print("     _");
+    tty->print("   _");
+  } else {
+    print_node_idx(b->head());
```

Comment:
I think it would also be useful to print the block identifier, i.e. `print("B%d", b->_pre_order)`.